java-checks/src/main/java/org/sonar/java/checks/OneExpectedCheckExceptionCheck.java

```diff
@@ -0,0 +1,206 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionStatementTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
+
+@Rule(key = "S5783")
+public class OneExpectedCheckExceptionCheck extends IssuableSubscriptionVisitor {
+
+  private static final String JUNIT4_ASSERT = "org.junit.Assert";
+
+  private static final MethodMatchers JUNIT4_ASSERT_THROWS_WITH_MESSAGE = MethodMatchers.create()
+    .ofTypes(JUNIT4_ASSERT)
+    .names("assertThrows")
+    .addParametersMatcher("java.lang.String", MethodMatchers.ANY, MethodMatchers.ANY)
+    .build();
+
+  private static final MethodMatchers ALL_ASSERT_THROWS_MATCHER = MethodMatchers.create()
+    .ofTypes(JUNIT4_ASSERT, "org.junit.jupiter.api.Assertions")
+    .names("assertThrows")
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers JUNIT_FAIL_MATCHER = MethodMatchers.create()
+    .ofTypes(JUNIT4_ASSERT, "org.junit.jupiter.api.Assertions")
+    .names("fail")
+    .withAnyParameters()
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.TRY_STATEMENT, Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
```

Comment:
to be honest, I'm not sure this is still required in checks. From what I remember, we don't have situations anymore where semantic is not available.